### DISCUSSION TITLE
JSON Field Expressions in v3

### DISCUSSION THREAD

**Discussion by adamfisher** [elsa-workflows/elsa-core#6184] [Q&A] [answered]
In v2, we had the ability to specify JSON for the field type and in v3 this is no longer present.

![image](https://github.com/user-attachments/assets/57e50100-07e8-43ec-bb0b-a568f5670bd4)

v3 shows has options for field expressions listed below but none for JSON:

![image](https://github.com/user-attachments/assets/c2f78373-2dbf-4ff0-b47d-2ac255af0cf1)

No matter what variation I try:

```
return ["00.Comms", "01.Sales"]; // JavaScript field
return JSON.stringify(["00.Comms", "01.Sales"]); // JavaScript field
return JSON.parse(["00.Comms", "01.Sales"]); // JavaScript field
```

I get a similar error regarding deserialization no matter what I try:

```shell
Failed to deserialize ["00.Comms","01.Sales"] to Newtonsoft.Json.Linq.JArray
```

Here is how the field is defined in my activity:

```csharp
[Input(
	DisplayName = "Sales Folder Names",
	UIHint = InputUIHints.JsonEditor
)]
public Input<JArray?> SalesFolderNames { get; set; } = default!;
```

Even better if I could specify a strongly-typed model object directly for the field instead of a `JArray`? I only did this because v2 didn't seem to let me do it.

**Comment by Nekrosis1** [comment, marked as answer]
Elsa variables, inputs and outputs are generally ExpandoObjects behind the scenes, so instead of passing in the array as stringified, which results in a string, you might want to try to pass in your list as 
`return ["00.Comms", "01.Sales"]; `
and use
`public Input<List<string>> SalesFolderNames { get; set; } = default!;`
as your input in your activity.
if you wish to pass in more complex objects, which you have a model for, you can define the input as
`public Input<object> SalesFolderNames { get; set; } = default!;`
or
`public Input<List<object>> SalesFolderNames { get; set; } = default!;`
you can then do something like
```
if (context.Get(SalesFolderNames)  is SalesFolderModel salesFolderModel)
{
...
}
```
If getting a JArray is important, you could pass it as JSON.stringify() and then use `Input<string>`  and tryParse the string into JArray in your activity.

Not sure if I understood the problem correctly, hope this helps.

**Reply by adamfisher** [reply]
I modified the code to be `Input<List<string>>` and `Input<List<MyCustomObject>>` to match the actual types I needed and that worked! In v2 it was problematic because it wouldn't deserialize to my types and I had to settle for `JArray` and then convert them to the type I needed so this is a nice improvement. Hopefully as v3 roadmap progresses, the dynamic code completion and field validation will improve in Elsa Studio because that seems to be where I spend a good chunk of time trying to figure out what's going on with these types especially now since there is no JSON field type.
